File: ruby-elasticemail-examples/examples/ee.rb
# frozen_string_literal: true

##
# Shared setup for the standalone examples: loads .env, configures the SDK
# and exposes the env values every script needs.
#
# Usage in a script:
#   require_relative "ee"

require "bundler/setup"
require "ElasticEmail"
require "dotenv/load"
require "json"

API_KEY = ENV["ELASTICEMAIL_API_KEY"]
if API_KEY.nil? || API_KEY.empty?
  warn "ELASTICEMAIL_API_KEY is not set. Copy .env.example to .env and add your API key."
  exit 1
end

ElasticEmail.configure do |config|
  config.api_key["X-ElasticEmail-ApiKey"] = API_KEY
end

FROM = ENV.fetch("EMAIL_FROM", "Acme <hello@yourdomain.com>")
TO = ENV.fetch("EMAIL_TO", "you@yourdomain.com")
CONTACT_EMAIL = ENV.fetch("CONTACT_EMAIL", FROM)
LIST_NAME = ENV.fetch("ELASTICEMAIL_LIST_NAME", "Newsletter")
TEMPLATE_NAME = ENV.fetch("ELASTICEMAIL_TEMPLATE_NAME", "welcome-example")
PUBLIC_URL = ENV.fetch("PUBLIC_URL", "http://localhost:3000")
WEBHOOK_TOKEN = ENV.fetch("ELASTICEMAIL_WEBHOOK_TOKEN", "change_me")
SENDING_DOMAIN = ENV.fetch("SENDING_DOMAIN", "yourdomain.com")

# Prints the HTTP status and the API error body ({"Error": "..."}), then exits.
def api_error(step, e)
  warn "Error (#{step}): #{e.code} #{e.response_body || e.message}"
  exit 1
end

# Elastic Email answers 400 with an "already exists" message when a list or
# domain is created twice.
def already_exists?(e)
  e.code == 400 && e.response_body.to_s.match?(/exist|already/i)
end

File: kotlin-elasticemail-examples/src/main/kotlin/com/elasticemail/examples/Ee.kt
package com.elasticemail.examples

import com.elasticemail.client.ApiClient
import com.elasticemail.client.ApiException
import com.elasticemail.client.Configuration
import com.elasticemail.client.auth.ApiKeyAuth
import io.github.cdimascio.dotenv.Dotenv
import kotlin.system.exitProcess

/**
 * Shared setup for the standalone examples: loads .env, builds an authenticated ApiClient
 * and exposes the common environment values.
 */
object Ee {
    private val dotenv: Dotenv = Dotenv.configure().ignoreIfMissing().load()

    fun env(name: String, defaultValue: String): String = env(name) ?: defaultValue

    fun env(name: String): String? = dotenv.get(name)?.takeIf { it.isNotEmpty() }

    fun from(): String = env("EMAIL_FROM", "Acme <hello@yourdomain.com>")

    fun to(): String = env("EMAIL_TO", "you@yourdomain.com")

    fun client(): ApiClient {
        val apiKey = env("ELASTICEMAIL_API_KEY")
        if (apiKey == null) {
            System.err.println("ELASTICEMAIL_API_KEY environment variable is required")
            exitProcess(1)
        }

        val client = Configuration.getDefaultApiClient()
        (client.getAuthentication("apikey") as ApiKeyAuth).apiKey = apiKey
        return client
    }

    /** Prints the HTTP status and the API error body, e.g. {"Error":"..."}. */
    fun printApiError(step: String, e: ApiException) {
        System.err.println("Error ($step): ${e.code} ${errorBody(e)}")
    }

    fun errorBody(e: ApiException): String =
        e.responseBody?.takeIf { it.isNotEmpty() } ?: e.message.orEmpty()

    /** Prints the API error and exits with status 1. */
    fun fail(step: String, e: ApiException): Nothing {
        printApiError(step, e)
        exitProcess(1)
    }
}
